File: app/src/main/java/one/mixin/android/widget/markdown/handler/BoldEditHandler.java
package one.mixin.android.widget.markdown.handler;

import android.text.Editable;
import android.text.Spanned;
import android.text.TextPaint;
import android.text.style.MetricAffectingSpan;
import androidx.annotation.NonNull;
import io.noties.markwon.core.spans.StrongEmphasisSpan;
import io.noties.markwon.editor.AbstractEditHandler;
import io.noties.markwon.editor.MarkwonEditorUtils;
import io.noties.markwon.editor.PersistedSpans;

public class BoldEditHandler extends AbstractEditHandler<StrongEmphasisSpan> {
    @Override
    public void configurePersistedSpans(@NonNull PersistedSpans.Builder builder) {
        // Here we define which span is _persisted_ in EditText, it is not removed
        //  from EditText between text changes, but instead - reused (by changing
        //  position). Consider it as a cache for spans. We could use `StrongEmphasisSpan`
        //  here also, but I chose Bold to indicate that this span is not the same
        //  as in off-screen rendered markdown
        builder.persistSpan(Bold.class, Bold::new);
    }

    @Override
    public void handleMarkdownSpan(
            @NonNull PersistedSpans persistedSpans,
            @NonNull Editable editable,
            @NonNull String input,
            @NonNull StrongEmphasisSpan span,
            int spanStart,
            int spanTextLength) {
        // Unfortunately we cannot hardcode delimiters length here (aka spanTextLength + 4)
        //  because multiple inline markdown nodes can refer to the same text.
        //  For example, `**_~~hey~~_**` - we will receive `**_~~` in this method,
        //  and thus will have to manually find actual position in raw user input
        final MarkwonEditorUtils.Match match =
                MarkwonEditorUtils.findDelimited(input, spanStart, "**", "__");
        if (match != null) {
            editable.setSpan(
                    // we handle StrongEmphasisSpan and represent it with Bold in EditText
                    //  we still could use StrongEmphasisSpan, but it must be accessed
                    //  via persistedSpans
                    persistedSpans.get(Bold.class),
                    match.start(),
                    match.end(),
                    Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
            );
        }
    }

    @NonNull
    @Override
    public Class<StrongEmphasisSpan> markdownSpanType() {
        return StrongEmphasisSpan.class;
    }


    static class Bold extends MetricAffectingSpan {
        public Bold() {
            super();
        }

        @Override
        public void updateDrawState(TextPaint tp) {
            update(tp);
        }

        @Override
        public void updateMeasureState(@NonNull TextPaint textPaint) {
            update(textPaint);
        }

        private void update(@NonNull TextPaint paint) {
            paint.setFakeBoldText(true);
        }
    }
}

File: app/src/main/java/one/mixin/android/widget/markdown/handler/BoldEditHandler.kt
package one.mixin.android.widget.markdown.handler

import android.text.Editable
import android.text.Spanned
import android.text.TextPaint
import android.text.style.MetricAffectingSpan
import io.noties.markwon.core.spans.StrongEmphasisSpan
import io.noties.markwon.editor.AbstractEditHandler
import io.noties.markwon.editor.MarkwonEditorUtils
import io.noties.markwon.editor.PersistedSpans

class BoldEditHandler : AbstractEditHandler<StrongEmphasisSpan>() {
    override fun configurePersistedSpans(builder: PersistedSpans.Builder) {
        builder.persistSpan(Bold::class.java) { Bold() }
    }

    override fun handleMarkdownSpan(
        persistedSpans: PersistedSpans,
        editable: Editable,
        input: String,
        span: StrongEmphasisSpan,
        spanStart: Int,
        spanTextLength: Int
    ) {
        val match = MarkwonEditorUtils.findDelimited(input, spanStart, "**", "__")
        if (match != null) {
            editable.setSpan(
                persistedSpans.get(Bold::class.java),
                match.start(),
                match.end(),
                Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
            )
        }
    }

    override fun markdownSpanType(): Class<StrongEmphasisSpan> {
        return StrongEmphasisSpan::class.java
    }

    internal class Bold : MetricAffectingSpan() {
        override fun updateDrawState(tp: TextPaint) {
            update(tp)
        }

        override fun updateMeasureState(textPaint: TextPaint) {
            update(textPaint)
        }

        private fun update(paint: TextPaint) {
            paint.isFakeBoldText = true
        }
    }
}
